### DISCUSSION TITLE
Some cache keys dots are URI encoded

### DISCUSSION THREAD

**Discussion by PatrykWalach** [urql-graphql/urql#2528] [Q&A] [answered]
I have schema with embed entities
```gql 
type MediaCoverImage {
  medium
  large
}

type Media {
   id: Int 
   coverImage: MediaCoverImage
}
```
I've setup my graph cache with keys and my `requestPolicy` as `cache-first`
```ts
keys : {
  MediaCoverImage: () => null,
}
```
however after mutation, the queries miss the cache
I've used an `ofllineExchange` with `localStorage` as in the [documentation](https://formidable.com/open-source/urql/docs/graphcache/offline/#custom-storages) and noticed that the keys for embed data types looks like this: `Media:220832%2ecoverImage.medium`
`%2e` is URI encoded `.`
After changing the implementation of the storage to URI decode keys, I've started to hit the cache after mutations
```ts
  readData() {
    return Promise.resolve().then(() => {
      const local = localStorage.getItem('data') || null;
      Object.assign(cache, JSON.parse(local));
      return Object.fromEntries(
        Object.entries(cache).map(([key, value]) => [decodeURI(key), value])
      );
    });
},
```
Are the dots encoded intentional, am I missing cache because of something obvious?

**Comment by JoviDeCroock** [comment, marked as answer]
When you talk about missing cache I assume you mean rehydration, as we never read from this store during the normal running webapp. The data stored there in localStorage is _only_ used when the app boots up to put data in-memory, so just to be sure we are on the same wave-length here, when you say you are missing cache then you are talking about doing:

- boot app
- the relevant query misses

As a mutation would mean that your updater is doing something unexpected and wouldn't refer to the localStorage cache as post-mutation you won't be hitting said cache

**Reply by PatrykWalach** [reply]
After your response I've noticed that the problem is somewhere else, so I'll try to elaborate on my setup.

I'm using new `react-router` where each route has a loader (in my case call to `client.query`) and action (in my case call to `client.mutate`). The router reexecutes loaders after every action.
What I'm trying to do is:
 - boot app
 - router fetches - `client.query` does an online request
 - go offline
 - perform an action - `client.mutate` does optimistic update fails an online request but keeps an optimistic update
 - router reexecutes - this time `client.query` doesn't make a request, just reads from cache

I thought using `offlineExchange` was helping since there was not online requests, but it was actually causing loaders to not reexecute at all, displaying stale data.

At the moment I have few loaders, all of these reexecute: 
- one for Viewer - this one hits the cache
- second where I execute below query for every day of the week - I get seven online requests after mutation
  ```gql
  query SchedulesQuery($airingAt_greater: Int, $airingAt_lesser: Int, $page: Int) {
    Page(page: $page) {
      airingSchedules(airingAt_greater: $airingAt_greater, airingAt_lesser: $airingAt_lesser) {
        media {
          id
          mediaListEntry{
            id
            progress
          }
        }
      }
    }
  }
  ```

This is my mutation, I use it to increment `MediaList.progress`
```gql
  mutation MediaProgressMutation($mediaId: Int!, $progress: Int!) {
    SaveMediaListEntry(mediaId: $mediaId, progress: $progress) {
      id
      progress
    }
  }
```

And cache, initially `Media.mediaListEntry` is `null` so I try to link it in `updates`
```ts
updates: {
    Mutation: {
      SaveMediaListEntry(result, args, cache) {
        if (args.mediaId) {
          const media = cache.keyOfEntity({
            __typename: 'Media',
            id: args.mediaId
          });

          const mediaList = cache.keyOfEntity(result.SaveMediaListEntry);
          cache.link(media, 'mediaListEntry', mediaList);
        }
      }
    }
  },
  optimistic: {
    SaveMediaListEntry(args) {
      return {
        __typename: 'MediaList',
        id: Math.random(),
        ...args
      };
    }
  },
```

**Reply by JoviDeCroock** [reply]
If you execute this flow

- query
- mutation

Without the offlineExchange does your optimistic update work then, if it does we can start narrowing down the issue 😄 basically what I am trying to find out is whether your issue is located in the updater or not

Essentially what in your case happens behind the scenes is

- you have queried data
- you execute a mutation
- we perform the optimistic update (and the query is automatically updated with the result if it's a useQuery)
- we see that you are offline
- we store the mutation in localStorage to perform it again when you get back online

**Reply by PatrykWalach** [reply]
Hey, thanks for your time, you helped my out a lot, the issue was in the optimistic response, I was overriding the id.
Have a great day!
```ts
  optimistic: {
    SaveMediaListEntry(args, cache) {
      const mediaEntry = cache.resolve(
        {
          __typename: 'Media',
          id: args.mediaId!
        },
        'mediaListEntry'
      );

      const id = cache.resolve(mediaEntry, 'id') ?? Math.random();

      return {
        __typename: 'MediaList',
        id,
        ...args
      };
    }
  },
  ```
